lib/ddtrace/transport/traces.rb

```diff
@@ -8,25 +9,177 @@ module Traces
+        DEFAULT_MAX_PAYLOAD_SIZE = 5 * 1024 * 1024
+
+        attr_reader :encoder, :max_size
+
+        #
+        # Single traces larger than +max_size+ will be discarded.
+        #
+        # @param encoder [Datadog::Encoding::Encoder]
+        # @param max_size [String] maximum acceptable payload size
+        def initialize(encoder, max_size: DEFAULT_MAX_PAYLOAD_SIZE)
+          @encoder = encoder
+          @max_size = max_size
+        end
+
+        # Encodes a list of traces in chunks.
+        # Before serializing, all traces are normalized. Trace nesting is not changed.
+        #
+        # @param traces [Enumerable<Trace>] list of traces
+        # @return [Enumerable[Array[Bytes,Integer]]] list of encoded chunks: each containing a byte array and
+        #   number of traces
+        def encode_in_chunks(traces)
+          encoded_traces = traces.map { |t| encode_one(t) }.reject(&:nil?)
+
+          Datadog::Chunker.chunk_by_size(encoded_traces, max_size).map do |chunk|
+            [encoder.join(chunk), chunk.size]
+          end
+        end
+
+        private
+
+        def encode_one(trace)
+          encoded = Encoder.encode_trace(encoder, trace)
+
+          if encoded.size > max_size
+            # This single trace is too large, we can't flush it
+            Datadog.logger.debug { "Dropping trace. Payload too large: '#{trace.map(&:to_hash)}'" }
+            Datadog.health_metrics.transport_trace_too_large(1, tags: ["max_size:#{max_size}"])
+
+            return nil
+          end
+
+          encoded
+        end
+      end
+
+      # Encodes traces using {Datadog::Encoding::Encoder} instances.
+      module Encoder
+        module_function
+
+        def encode_trace(encoder, trace)
+          encoder.encode(trace.map(&:to_hash))
+        end
+      end
+
+      # Sends traces based on transport API configuration.
+      #
+      # This class initializes the HTTP client, breaks down large
+      # batches of traces into smaller chunks and handles
+      # API version downgrade handshake.
+      class Transport
+        attr_reader :client, :apis, :default_api, :current_api_id
+
+        def initialize(apis, default_api)
+          @apis = apis
+          @default_api = default_api
+
+          change_api!(default_api)
+        end
+
+        def send_traces(traces)
+          encoder = current_api.encoder
+          chunker = Datadog::Transport::Traces::Chunker.new(encoder)
+
+          responses = chunker.encode_in_chunks(traces.lazy).map do |encoded_traces, trace_count|
+            request = Datadog::Transport::Traces::Request.new(encoded_traces, trace_count)
+
+            client.send_payload(request).tap do |response|
+              if downgrade?(response)
+                downgrade!
+                return send_traces(traces.lazy)
+              end
+            end
+          end.force
+
+          Datadog.health_metrics.transport_chunked(responses.size, tags: ["max_size:#{chunker.max_size}"])
```

Comment:
What's the difference between `size` and `max_size` and what does it tell us? Just want to know what benefit this extra tag provides.

Reply:
In case the user wants to tweak the `max_size`, they can track how that affects the numbers of chunks generated in their application.
It's a long-shot, tbh, I'm not strongly attached to this tag.

Reply:
If the user tweaks the max size and we tag it, do you have a use case in mind for how we'd use this information?

I think if we can say "it'll help us determine whether chunks are too big when nothing else will", or there's some other articulated case for how we use the tag, I think the tag is probably fine to add. Otherwise probably better to omit. I'm on the fence.

Reply:
please do not tag with max size. This is not a good tag. The cardinality on it is unbounded, and its usability in dashboards/monitors won't be any good.

Instead consider creating separate metrics for number of responses, and the number of requests sent/chunks created/etc.

Reply:
It should be always the same value today for every client, and it's fixed once the application starts, but I've removed it as per recommendation.